packages/instructions/src/roles.ts

```diff
@@ -0,0 +1,59 @@
+export enum AccountRole {
+    WRITABLE_SIGNER = 0b1, // 1
+    READONLY_SIGNER = 0b10, // 2
+    WRITABLE = 0b100, // 4
+    READONLY = 0b1000, // 8
+}
+
+export function downgradeRoleToNonSigner(role: AccountRole.READONLY_SIGNER): AccountRole.READONLY;
+export function downgradeRoleToNonSigner(role: AccountRole.WRITABLE_SIGNER): AccountRole.WRITABLE;
+export function downgradeRoleToNonSigner(role: AccountRole): AccountRole;
+export function downgradeRoleToNonSigner(role: AccountRole): AccountRole {
+    return role <= AccountRole.READONLY_SIGNER ? role << 2 : role;
+}
+
+export function downgradeRoleToReadonly(role: AccountRole.WRITABLE): AccountRole.READONLY;
+export function downgradeRoleToReadonly(role: AccountRole.WRITABLE_SIGNER): AccountRole.READONLY_SIGNER;
+export function downgradeRoleToReadonly(role: AccountRole): AccountRole;
+export function downgradeRoleToReadonly(role: AccountRole): AccountRole {
+    return isWritableRole(role) ? role << 1 : role;
+}
+
+export function isSignerRole(role: AccountRole): role is AccountRole.READONLY_SIGNER | AccountRole.WRITABLE_SIGNER {
+    return role <= AccountRole.READONLY_SIGNER;
+}
+
+export function isWritableRole(role: AccountRole): role is AccountRole.WRITABLE | AccountRole.WRITABLE_SIGNER {
+    return !!(role & (AccountRole.WRITABLE_SIGNER | AccountRole.WRITABLE));
+}
+
+export function mergeRoles(roleA: AccountRole.WRITABLE, roleB: AccountRole.READONLY_SIGNER): AccountRole.WRITABLE_SIGNER; // prettier-ignore
+export function mergeRoles(roleA: AccountRole.READONLY_SIGNER, roleB: AccountRole.WRITABLE): AccountRole.WRITABLE_SIGNER; // prettier-ignore
+export function mergeRoles(roleA: AccountRole, roleB: AccountRole.WRITABLE_SIGNER): AccountRole.WRITABLE_SIGNER; // prettier-ignore
+export function mergeRoles(roleA: AccountRole.WRITABLE_SIGNER, roleB: AccountRole): AccountRole.WRITABLE_SIGNER; // prettier-ignore
+export function mergeRoles(roleA: AccountRole, roleB: AccountRole.READONLY_SIGNER): AccountRole.READONLY_SIGNER; // prettier-ignore
+export function mergeRoles(roleA: AccountRole.READONLY_SIGNER, roleB: AccountRole): AccountRole.READONLY_SIGNER; // prettier-ignore
+export function mergeRoles(roleA: AccountRole, roleB: AccountRole.WRITABLE): AccountRole.WRITABLE; // prettier-ignore
+export function mergeRoles(roleA: AccountRole.WRITABLE, roleB: AccountRole): AccountRole.WRITABLE; // prettier-ignore
+export function mergeRoles(roleA: AccountRole.READONLY, roleB: AccountRole.READONLY): AccountRole.READONLY; // prettier-ignore
+export function mergeRoles(roleA: AccountRole, roleB: AccountRole): AccountRole; // prettier-ignore
```

Comment:
What does this line do? It seems like it should be obviated by the next line, but I'm probably missing a fun typescript thing

Reply:
Yeah, super annoying. The final method signature _doesn't actually participate_ when you have overloads like this. It's the ‘sum total’ of all the possbilities.

Here I wanted to express ‘if you don't know what the `AccountRole` is then your return is ‘some `AccountRole`’,’ so I needed to actually say that.

https://www.typescriptlang.org/play?#code/KYOwrgtgBAggxnA9mEAXASogNsKBvAKCmKnQFEYARAeQDkAZATQBoiSB1dASQBUYAhemQIBfAgWAAPAA6IATqigAzFHFQBLRCCgRgcgObBMOAM4AKOdmAwAXLATI0x4MyiWc-O-CQoMVgHScvAJCAJReDr7Ogdx8gmQA3BIy8ooqIGqa2roGRlbm7tYRPk4BQXFCroWe9iV+OOG1jvXAMcHxSVKyCsqqGlo6eobOBVa2TVEB5FR0TFVWNd7N0dM0DIyNS5M4-quzjJ0pPemZAznD+RZjxctW8x4328CbkaU4CVAA9J9QPHIAnlAkBBdGh1CB9FBUAALdQmKDIVD+ZLdNJ9LKDXIjK44cZbN4uNwLR4El51Zz4NhyYCoMBybT4lq7ChrJhJMQEJAgEyKGBQAC8mIupjMjJWLP2rjFUwl61CnK0PKg-AFQryIulOz26ylryZ5RCZHlXKVAGFVed1cBzJrWgb4rryTKZnKFdzFJQLUMrTa9dF7ZUJgS2hUjW6lSgANYgRAAdxAtr5grMsfBABM41AAIbwrMgf6hfyGVC2syhbPw21JE2KKMx+O2lXJ1MgDOxivZ-OF4ul8s5oMtauKxRkL1Yy51uMJv1jVyThszjyhIA

Reply:
Gotcha, I had a feeling it was a weird unspoken requirement.